File: CSharp/CubeAD/Side.cs
namespace CubeAD
{
	public struct Side
	{
		public const int FACE_COUNT = 8;
		public const int BIT_COUNT = FACE_COUNT * 3;
		public const int BIT_MASK = (1 << 24) - 1;

		public static uint[] STRIPE_MASK =
		{
			511U,
			511U << 6,
			511U << 12,
			7U | (63U << 18)
		};

		public static uint[] SQUARE_MASK =
{
			511U << 3,
			511U << 9,
			511U << 15,
			63U | (7U << 21)
		};

		uint Data;

		public uint this[int i]
		{
			get
			{
				return (Data >> (i * 3)) & 7;
			}
			set
			{
				Data &= ~(7U << i * 3);
				Data |= value << (i * 3);
			}
		}

		public CubeColor this[int x, int y]
		{
			get
			{
				if (y == 0) return (CubeColor)this[x];
				if (y == 2) return (CubeColor)this[6 - x];

				return (CubeColor)this[x == 0 ? 7 : 3];
			}
		}

		public Side(uint val)
		{
			Data = val;
		}

		public Side(CubeColor color)
		{
			uint val = (uint)color;
			Data = 0;
			for (int i = 0; i < 8; i++)
			{
				this[i] = val;
			}
		}

		public bool FitsPattern(Side color, uint bitmask)
		{
			return ((Data ^ color.Data) & bitmask) == 0;
		}
		public bool IsPair(int index)
		{
			return this[index] == this[(index + 1) % 8];
		}

		public static uint GenerateBitmask(uint compressed)
		{
			return ((compressed & 1) * 7) |
				((compressed & 2) * (7 << 2)) |
				((compressed & 4) * (7 << 4)) |
				((compressed & 8) * (7 << 6)) |
				((compressed & 16) * (7 << 8)) |
				((compressed & 32) * (7 << 10)) |
				((compressed & 64) * (7 << 12)) |
				((compressed & 128) * (7 << 14));
		}

		public void Rotate(int count)
		{
			Data = ((Data << (count * 6)) & BIT_MASK) | (Data >> (BIT_COUNT - count * 6));
		}
		public static uint RotateVal(uint val, int count)
		{
			return ((val << (count * 6))) & BIT_MASK | (val >> (BIT_COUNT - count * 6));
		}

		public uint GetStripe(int stripe)
		{
			return RotateVal(Data & STRIPE_MASK[stripe], (4 - stripe) & 3);
		}
		public void SetStripe(int stripe, uint val)
		{
			Data &= ~STRIPE_MASK[stripe];
			Data |= RotateVal(val, stripe);
		}

		public static bool operator ==(Side a, Side b)
		{
			return a.Data == b.Data;
		}
		public static bool operator !=(Side a, Side b)
		{
			return a.Data != b.Data;
		}

		public string WriteAsSide(char center)
		{
			string s = "";
			for (int i = 0; i < 3; i++)
				s += ((CubeColor)this[i]).ToString()[0] + " ";

			s += "\n";
			s += ((CubeColor)this[7]).ToString()[0] + " ";
			s += center + " ";
			s += ((CubeColor)this[3]).ToString()[0] + "\n";

			for (int i = 3 - 1; i >= 0; i--)
				s += ((CubeColor)this[i + 4]).ToString()[0] + " ";

			return s;
		}

		public override string ToString()
		{
			string s = "";
			for (int i = 0; i < 8; i++)
			{
				s += this[i] + " ";
			}

			return s;
		}

		public override bool Equals(object obj)
		{
			return this == (Side)obj;
		}

		public override int GetHashCode()
		{
			return (int)Data;
		}
	}
}

File: Rost/src/side.rs
use crate::enums::{CubeColor, u8_to_CubeColor};

//consts
pub const FACE_COUNT: i32 = 8;
pub const BIT_COUNT: i32 = FACE_COUNT * 3;
pub const BIT_MASK: u32 = (1 << 24) - 1;

pub const  STRIPE_MASK: [u32; 4] = [
    511_u32,
    511_u32 << 6,
    511_u32 << 12,
    7_u32 | (63_u32 << 18)];

pub const SQUARE_MASK: [u32; 4] = [
    511_u32 << 3,
    511_u32 << 9,
    511_u32 << 15,
    63_u32 | (7_u32 << 21)];

pub fn rotate_value(value: u32, count: i32) -> u32 {
    ((value << (count * 6)) & BIT_MASK) | (value >> (BIT_COUNT - count * 6))
}

pub const SOLVED_SIDES: [Side; 6] = [
    Side {data: 0},
    Side {data: 1 | (1 << 3) | (1 << 6) | (1 << 9) | (1 << 12) | (1 << 15) | (1 << 18) | (1 << 21) },
    Side {data: 2 | (2 << 3) | (2 << 6) | (2 << 9) | (2 << 12) | (2 << 15) | (2 << 18) | (2 << 21) },
    Side {data: 3 | (3 << 3) | (3 << 6) | (3 << 9) | (3 << 12) | (3 << 15) | (3 << 18) | (3 << 21) },
    Side {data: 4 | (4 << 3) | (4 << 6) | (4 << 9) | (4 << 12) | (4 << 15) | (4 << 18) | (4 << 21) },
    Side {data: 5 | (5 << 3) | (5 << 6) | (5 << 9) | (5 << 12) | (5 << 15) | (5 << 18) | (5 << 21) },
];


#[derive(Eq, PartialEq, Hash, Clone, Copy)]
pub struct Side {
    data: u32,
}

impl Side {
    pub fn new(value: u32) -> Self {
        Side { data: value }
    }    

    pub fn get_color(&self, index: i32) -> CubeColor {
            u8_to_CubeColor(((self.data >> (index * 3)) & 7) as u8)
    }
    pub fn get_color_2d(&self, x: i32, y: i32) -> CubeColor {
        if y == 0 {
            return self.get_color(x);
        }

        if y == 2 {
            return self.get_color(6 - x)
        }
        
        return self.get_color(if x == 0 {7} else {3});
    }

    pub fn set_color(&mut self, index: i32, color: CubeColor) {        
        self.data &= !(7_u32 << index * 3);
        self.data |= (color as u32) << (index * 3);
    }

    pub fn get(&self, index: i32) -> u32 {
        (self.data >> (index * 3)) & 7
    }
    pub fn set(&mut self, index: i32, color: u32) {
        self.data &= !(7_u32 << index * 3);
        self.data |= color << (index * 3);
    }

    pub fn fits_pattern(&self, pattern: Side, bitmask: u32) -> bool {
        (self.data ^ pattern.data) & bitmask == 0_u32
    }

    pub fn is_pair(&self, index: i32) -> bool{
        self.get(index) == self.get((index + 1) % 8)
    }

    pub fn rotate(&mut self, count: i32){
        self.data = ((self.data << (count * 6)) & BIT_MASK) | (self.data >> (BIT_COUNT - count * 6));
    }

    pub fn get_stripe(&self, stripe: i32) -> u32 {
        rotate_value(self.data & STRIPE_MASK[stripe as usize], (4 - stripe) & 3)
    }

    pub fn set_stripe(&mut self, stripe: i32, value: u32){
        self.data &= !STRIPE_MASK[stripe as usize];
        self.data |= rotate_value(value, stripe);
    }

    pub fn print_raw(&self){
        for i in 0..8 {
            print!("{} ", self.get_color(i).to_string());
        }
    }
}